app/decorators/models/product/add_relation_methods.rb

```diff
@@ -8,7 +8,11 @@ def self.prepended(base)
 
         # When a Spree::Product is destroyed, we also want to destroy all Spree::Relations
         # "from" it as well as "to" it.
-        base.after_destroy :destroy_product_relations
+        if base.respond_to?(:after_discard)
```

Comment:
I think this should not be an `if/else` statement. I mean we should have both.

Currently [Solidus backend](https://github.com/solidusio/solidus/blob/f0ea6b0aa0a6c3ddbdeea87dadf9e6bee41442f0/backend/app/controllers/spree/admin/products_controller.rb#L47) removes products via `discard`, so `after_discard` is called. But when the product is destroyed via `destroy` instance method (in old gems, old applications, rails console... whatever), I think we still want to trigger the `after_destroy` hook. 

So the code should be IMO: 
```ruby
base.after_discard(:destroy_product_relations) if base.respond_to?(:after_discard)
base.after_destroy :destroy_product_relations
```
@kennyadsl @Draiken what do you think?

Reply:
I just realized @kennyadsl already did notice what I wrote in my previous comment... sorry for the noise!

@Draiken I think this PR only needs some more tests, please let me know if you need help in completing it, I'd be happy to help.

Reply:
Sorry, I should've finished this ages ago.
I'll change the logic and add the test today.